aten/src/ATen/native/cuda/EmbeddingBackwardKernel.cu

```diff
@@ -0,0 +1,311 @@
+    ret[id] = ceil_div(size, blocksize);
+  }
+}
+
+__global__
+void split_segment_offsets_kernel(
+        int64_t *ret,
+        const int64_t *segment_sizes,
+        const int64_t *segment_sizes_offsets,
+        const int64_t *segment_offsets,
+        int64_t num_of_segments,
+        int64_t blocksize) {
+  const int id = blockIdx.x * blockDim.x + threadIdx.x;
+  if(id < num_of_segments) {
+    int64_t idx = segment_sizes_offsets[id];
+    const int64_t segment_size = segment_sizes[id];
+    const int64_t segment_offset = segment_offsets[id];
+    for (int64_t i=0; i<segment_size; ++i) {
+      ret[idx++] = segment_offset + i * blocksize;
+    }
+  }
+}
+
+
+template <typename scalar_t>
+__global__ void compute_grad_weight_bags(
+    int64_t *indices, scalar_t *gradOutput,
+    int64_t *offset2bag, int64_t *count, ptrdiff_t numel,
+    int64_t stride, int mode_mean, const int64_t *bag_size,
+    scalar_t* per_sample_weights, int64_t per_sample_weights_stride,
+    int64_t* segment_offsets, int64_t num_of_segments, scalar_t *grad_weight_per_segment,
+    const int64_t stride_warped) {
+
+  const int gid = blockIdx.x * blockDim.x + threadIdx.x;
+  const int id = gid / stride_warped;
+  const int startFeature = gid % stride_warped;
+  if (startFeature >= stride) {
+    return;
+  }
+  if (id >= num_of_segments) {
+    return;
+  }
+  const int idx_begin = segment_offsets[id];
+  const int idx_end = (id == num_of_segments-1)?numel:segment_offsets[id+1];
+
+  acc_type<scalar_t, true> weight = 0;
+  for (int idx=idx_begin; idx < idx_end; ++idx) {
+    const int origRow = indices[idx];
+    const int seq_number = offset2bag[origRow];
+    const int gradOutputRow = seq_number * stride;
+
+    acc_type<scalar_t, true> scale = count ? 1.0 / count[idx] : 1.0;
+    if (per_sample_weights) {
+      scale *= per_sample_weights[origRow * per_sample_weights_stride];
+    }
+
+    acc_type<scalar_t, true> gradient = gradOutput[gradOutputRow + startFeature];
+    if (mode_mean) {
+      gradient /= bag_size[seq_number];
+    }
+    weight += gradient * scale;
+  }
+  grad_weight_per_segment[id * stride + startFeature] = weight;
+}
+
+template <typename scalar_t>
+__global__ void compute_grad_weight(
+    int64_t *indices,
+    scalar_t *gradOutput,
+    int64_t *count,
+    ptrdiff_t numel,
+    int64_t stride,
+    int64_t* segment_offsets,
+    int64_t num_of_segments,
+    scalar_t *grad_weight_per_segment,
+    int padding_idx,
+    const int64_t stride_warped) {
+
+  using accscalar_t = acc_type<scalar_t, true>;
+  const int gid = blockIdx.x * blockDim.x + threadIdx.x;
+  const int id = gid / stride_warped;
+  const int startFeature = gid % stride_warped;
+  if (startFeature >= stride) {
+    return;
+  }
+  if (id >= num_of_segments) {
+    return;
+  }
+  const int idx_begin = segment_offsets[id];
+  const int idx_end = (id == num_of_segments-1)?numel:segment_offsets[id+1];
+  if (idx_begin == padding_idx) {
```

Comment:
Isn't `padding_idx` defined based on the embedding index instead of the index of indices array? It seems to me `padding_idx` needs to be compared with `original_indices[x]` instead of `idx_begin`.

Reply:
it is compared with indices[idx_begin] below, so this indeed seems wrong or deserves a comment